### DISCUSSION TITLE
Browser testing with Servo

### DISCUSSION THREAD

**Discussion by jsudelko** [servo/servo#32826] [Ideas]
Given that Servo will not see mass market adoption for some time, you should find a niche to grow your user base. I suggest doubling down on browser testing for the JS ecosystem. The state of browser testing is pretty bad today. It's a market segment with well defined challenges, and neither a clear winner nor particularly steep competition.

DOM implementations in JS do not implement styling/layout. [HappyDOM ](https://github.com/capricorn86/happy-dom) is fast, but not entirely spec compliant. [JSDOM](https://github.com/jsdom/jsdom) is spec compliant but comparatively slow.

[Playwright](https://github.com/microsoft/playwright), [Puppeteer](https://github.com/puppeteer/puppeteer) and friends allow testing in Real Browsers™, but using them is like trying to run a marathon with an elephant tied to your waist. The ease of use is [a B- at best](https://developer.chrome.com/blog/webdriver-bidi-2023/).

Make every part of Servo scriptable and with simple straight forward APIs. Do things other browsers can't. e.g. use something like [napi-rs](https://github.com/napi-rs/napi-rs) to support testing Node.js, Bun and Deno. Approach it from first principles. We want bicycles not rocket ships. Bicycles not rocket ships.

**Comment by zvictor** [comment]
There is definitely demand for that! Specially these days that AI agents need to run browsers and all the options in the market suck really bad for that. Servo could be that missing piece.

**Comment by nalladev** [comment]
Does the android app collect any crash data for fixing issues? How do I report a broken site? If it automatically collects crash data then I just have to visit as much sites as possible providing you with data on bugs.

**Reply by jdm** [reply]
It does not. You can file issues about specific sites that do not work.

**Comment by ChALkeR** [comment]
I added Servo support to my [node:test](https://nodejs.org/api/test.html) and Jest-compatible [test runner](https://github.com/ExodusOSS/test#exodustest)

See https://github.com/ExodusOSS/bytes/pull/6/files

You can see it in action here on CI: https://github.com/ExodusOSS/bytes/actions/runs/20208778083/job/58011401615?pr=6

It doesn't target _browser testing_, just JS libs testing. i.e. no screenshots, devtools protocol, etc.
One could do whatever they want from inside the `test()` functions though

It supports things like snapshots and module mocks
